test/core/end2end/h2_ssl_certs_curves_test.cc

```diff
@@ -0,0 +1,370 @@
+#include "src/core/lib/security/credentials/ssl/ssl_credentials.h"
+#include "test/core/end2end/cq_verifier.h"
+#include "test/core/util/port.h"
+#include "test/core/util/test_config.h"
+
+constexpr char kCaCertPath0[] = "src/core/tsi/test_creds/ca_p256.pem";
+constexpr char kCaCertPath1[] = "src/core/tsi/test_creds/ca_p384.pem";
+constexpr char kCaCertPath2[] = "src/core/tsi/test_creds/ca_p521.pem";
+constexpr char kClientCertPath0[] = "src/core/tsi/test_creds/client_p256.pem";
+constexpr char kClientKeyPath0[] = "src/core/tsi/test_creds/client_p256.key";
+constexpr char kClientCertPath1[] = "src/core/tsi/test_creds/client_p384.pem";
+constexpr char kClientKeyPath1[] = "src/core/tsi/test_creds/client_p384.key";
+constexpr char kClientCertPath2[] = "src/core/tsi/test_creds/client_p521.pem";
+constexpr char kClientKeyPath2[] = "src/core/tsi/test_creds/client_p521.key";
+constexpr char kServerCertPath0[] = "src/core/tsi/test_creds/server1_p256.pem";
+constexpr char kServerKeyPath0[] = "src/core/tsi/test_creds/server1_p256.key";
+constexpr char kServerCertPath1[] = "src/core/tsi/test_creds/server1_p384.pem";
+constexpr char kServerKeyPath1[] = "src/core/tsi/test_creds/server1_p384.key";
+constexpr char kServerCertPath2[] = "src/core/tsi/test_creds/server1_p521.pem";
+constexpr char kServerKeyPath2[] = "src/core/tsi/test_creds/server1_p521.key";
+
+namespace grpc {
+namespace testing {
+namespace {
+
+gpr_timespec five_seconds_time() { return grpc_timeout_seconds_to_deadline(5); }
+
+grpc_server* server_create(grpc_completion_queue* cq, const char* server_addr,
+                           const char* ca_cert_path,
+                           const char* server_cert_path,
+                           const char* server_key_path) {
+  grpc_slice ca_slice, cert_slice, key_slice;
+  GPR_ASSERT(GRPC_LOG_IF_ERROR("load_file",
+                               grpc_load_file(ca_cert_path, 1, &ca_slice)));
+  GPR_ASSERT(GRPC_LOG_IF_ERROR(
+      "load_file", grpc_load_file(server_cert_path, 1, &cert_slice)));
+  GPR_ASSERT(GRPC_LOG_IF_ERROR("load_file",
+                               grpc_load_file(server_key_path, 1, &key_slice)));
+  const char* ca_cert =
+      reinterpret_cast<const char*> GRPC_SLICE_START_PTR(ca_slice);
+  const char* server_cert =
+      reinterpret_cast<const char*> GRPC_SLICE_START_PTR(cert_slice);
+  const char* server_key =
+      reinterpret_cast<const char*> GRPC_SLICE_START_PTR(key_slice);
+  grpc_ssl_pem_key_cert_pair pem_cert_key_pair = {server_key, server_cert};
+  auto* cert_config =
+      grpc_ssl_server_certificate_config_create(ca_cert, &pem_cert_key_pair, 1);
+  auto* options = grpc_ssl_server_credentials_create_options_using_config(
+      GRPC_SSL_REQUEST_CLIENT_CERTIFICATE_AND_VERIFY, cert_config);
+  grpc_server_credentials* server_creds =
+      grpc_ssl_server_credentials_create_with_options(options);
+  // This is a hack but we don't have a public API to force TLS version yet.
```

Comment:
Please explain in the test why we need to force TLS 1.2.

Reply:
Done.